cpp/src/reshape/byte_cast.cu

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/copying.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/replace.hpp>
+#include <cudf/reshape.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+namespace cudf {
+namespace {
+
+struct byte_list_conversion {
+  /**
+   * @brief Function object for converting primitive types and string columns to lists of bytes,
+   * mimics Spark's cast to binary type.
+   */
+  struct flip_endianness_lambda {
+    char* d_chars;
+    const char* d_data;
+    uint32_t mask;
+    __device__ void operator()(int byte_index)
+    {
+      d_chars[byte_index] = d_data[byte_index + mask - ((byte_index & mask) << 1)];
+    }
+  };
+
+  template <typename T>
+  struct normalize_lambda {
+    T* d_result;
+    const T* d_data;
+    __device__ void operator()(int index)
+    {
+      auto e = d_data[index];
+      if (isnan(e)) {
+        d_result[index] = std::numeric_limits<T>::quiet_NaN();
+      } else if (T{0.0} == e) {
+        d_result[index] = T{0.0};
+      } else
+        d_result[index] = e;
+    }
+  };
+
+  template <typename T>
+  struct normalize_flip_endianess_lambda {
```

Comment:
You shouldn't need to couple flipping endianess and normalization. All the floating point code path should so is tack on a transform iterator that performs the `normalize` function and just forwards to another function that works agnostically on iterators. Otherwise there is a bunch of redundancy between the integral and floating point code paths.